Feat: Set `migration.stateful=true` be default when creating a new VM (#12832)
canonical/lxd: doc/howto/move_instances.md

Virtual machines can be moved to another server while they are running, thus without any downtime.

To allow for live migration, you must enable support for stateful migration.

> To do so, ensure the following configuration if your instance is not on a shared storage backend:

Reviewer: Why the mention of shared storage here? This isn't related to size.state, the instances.migration.stateful setting is storage pool agnostic.

> To do so, ensure the following configuration are set if your instance is not on a shared storage backend:

Reviewer 2: ```suggestion
To do so, ensure the following configuration if your instance is not on a shared storage backend:
```

* Set {config:option}`instance-migration:migration.stateful` to `true` on the instance.
* Set [`size.state`](devices-disk) of the virtual machine's root disk device to at least the size of the virtual machine's {config:option}`instance-resource-limits:limits.memory` setting.

If you create an instance on a shared storage backend like Ceph RBD, you don't need to set any parameters to enable the live migration. When an instance is created on a shared storage backend, {config:option}`instance-migration:migration.stateful` is set to `true` by default for the instance. Additionally, the [`size.state`](devices-disk) parameter is unset, because it is not required for shared storage.

> The default value for {config:option}`instance-migration:migration.stateful` was introduced with LXD 5.21 and applies only to new instances. Existing instances won't be affected by this change.

Reviewer: We dont tend to mention versions in the docs, but rather leave that to the API extension functionality.

```{note}

> Some guest OSes (like Centos 8) do not support 9P file system share, which is used for instances that have stateful migration enabled.

Reviewer: This would be clearer if it expanded on the fact that LXD disables virtiofs filesystem shares when migration.stateful is enabled, and only exports them via 9p protocol. Thus if your guest OS doesn't support 9p the filesystem shares don't work with migration.stateful enabled. We should also mention that this will mean the lxd-agent won't work either.

* Set {config:option}`instance-migration:migration.stateful` to `true` on the instance.
* Set [`size.state`](devices-disk) of the virtual machine's root disk device to at least the size of the virtual machine's {config:option}`instance-resource-limits:limits.memory` setting.

```{note}

> If you are using a shared storage pool to back your instance like Ceph RBD, you don't need to set the [`size.state`](devices-disk) property.

Reviewer: > If you are using a shared storage pool to back your instance like Ceph RBD, you don't need to set [`size.state`](devices-disk) to perform live migration.

> * Set {config:option}`instance-migration:migration.stateful` to `true` on the instance.

Reviewer 2: Can you add to the intro of this list that you only need these settings for local storage backends?
It's always bad style to write "Do this now! Or actually, maybe you didn't need to ..." ;)

* Set [`size.state`](devices-disk) of the virtual machine's root disk device to at least the size of the virtual machine's {config:option}`instance-resource-limits:limits.memory` setting.

> However, if you create an instance on a shared storage backend like Ceph RBD, you don't have any parameters to set to enable the live migration. This is due to the fact that when the instance is created, it is by default set with {config:option}`instance-migration:migration.stateful` to `true` with the [`size.state`](devices-disk) parameter being unset a this is not strictly required for shared storage.

Reviewer 2: ```suggestion
If you create an instance on a shared storage backend like Ceph RBD, you don't need to set any parameters to enable the live migration. When an instance is created on a shared storage backend, {config:option}`instance-migration:migration.stateful` is set to `true` by default for the instance. Additionally, the [`size.state`](devices-disk) parameter is unset, because it is not required for shared storage.
```

Reviewer 2: We should probably also update the documentation for the default value of migration.stateful?

> This new default value for {config:option}`instance-migration:migration.stateful` only applies for new instances. Your existing instances won't be affected by this change.

Reviewer 2: ```suggestion
The default value for {config:option}`instance-migration:migration.stateful` was introduced with LXD 5.21 and applies only to new instances. Existing instances won't be affected by this change.
```

Remember that these are docs and not release notes, and nobody will know what "new" means. ;)

> Is is also important to note that, at the time of writing, some guest OSes like Centos 8, do not support 9P file system share (which is used for instance enabled with stateful migration) and therefore, file system disk shares won't work by default now.

Reviewer 2: ````suggestion
```{note}
Some guest OSes (like Centos 8) do not support 9P file system share, which is used for instances that have stateful migration enabled.
Therefore, file system disk shares won't work for such OSes unless you disable stateful migration.
```
````

* Set {config:option}`instance-migration:migration.stateful` to `true` on the instance.
* Set [`size.state`](devices-disk) of the virtual machine's root disk device to at least the size of the virtual machine's {config:option}`instance-resource-limits:limits.memory` setting.

```{note}

> If you are using a shared storage pool to back your instance like Ceph RBD, you don't need to set [`size.state`](devices-disk) to perform live migration.

Reviewer 2: ```suggestion
If you are using a shared storage pool like Ceph RBD to back your instance, you don't need to set [`size.state`](devices-disk) to perform live migration.
```

```

> The default value for {config:option}`instance-migration:migration.stateful` was introduced with the {config:option}`server-miscellaneous:instances.migration.stateful` introduced with the `instances_migration_stateful` API extension and applies only to new instances. Existing instances won't be affected by this change.

Reviewer 2: ```suggestion
The default value for {config:option}`instance-migration:migration.stateful` was introduced with the {config:option}`server-miscellaneous:instances.migration.stateful` option (which was introduced with the `instances_migration_stateful` API extension). It applies only to new instances. Existing instances won't be affected by this change.
```

If you are using a shared storage pool like Ceph RBD to back your instance, you don't need to set [`size.state`](devices-disk) to perform live migration.
```

> The default value for {config:option}`instance-migration:migration.stateful` was introduced with the {config:option}`server-miscellaneous:instances.migration.stateful` option (which was introduced with the `instances_migration_stateful` API extension). It applies only to new instances. Existing instances won't be affected by this change.

Reviewer: > It applies only to new instances. Existing instances won't be affected by this change.

This isn't true, please remove.

Reviewer 2: Then we can remove the whole paragraph.

Author: fixed

```{note}

> When {config:option}`instance-migration:migration.stateful` is enabled in LXD, virtiofs shares are disabled, and files are only shared via the 9P protocol. Consequently, guest OSes lacking 9P support, such as Centos 8, cannot share files with the host unless stateful migration is disabled. Additionally, the `lxd-agent` will not function for these guests under these conditions.

Reviewer: CentOS not Centos

If you are using a shared storage pool to back your instance like Ceph RBD, you don't need to set the [`size.state`](devices-disk) property.
```

> The default value for {config:option}`instance-migration:migration.stateful` was introduced with the {config:option}`server-miscellaneous:instances.migration.stateful` introduced with the `stateful_migration_default` API extension and applies only to new instances. Existing instances won't be affected by this change.

Reviewer: @ru-fu do we routinely mention API extensions in our general user docs, or is it assumed the docs for latest apply to the latest version of LXD only?

Reviewer 2: The docs apply only to the latest version of LXD. However, in this case, the latest version might be running instances that have been created earlier, so this information is relevant even for the latest version.
It's not something we do routinely, but imo we should point it out here.

Reviewer: thanks!

```{note}

> Because LXD disables virtiofs file system shares when {config:option}`instance-migration:migration.stateful` is enabled, and only exports them via 9p protocol, some guest OSes (like Centos 8) that do not support 9P file system share, won't be able to have a shared file system with the host (unless you disable stateful migration).

Reviewer: Can you try and make this a bit more concise without mentioning file system so often.
